test/src/test/java/hudson/model/UserTest.java

```diff
@@ -813,6 +813,25 @@ public void emptyUsernameConfigMigrated() {
         assertThat(empty.getFullName(), equalTo("Empty"));
     }
 
+    @Issue("JENKINS-47909")
+    @LocalData
+    @Test
+    public void shellyUsernameMigrated() {
+        File rootDir = new File(Jenkins.getInstance().getRootDir(), "users");
+        User user = User.getById("bla$phem.us", false);
+        assertCorrectConfig(user, "users/bla$0024phem.us/config.xml");
+        assertFalse(new File(rootDir, "bla$phem.us").exists());
+        assertTrue(user.getConfigFile().getFile().exists());
+        assertThat(user.getFullName(), equalTo("Weird Username"));
+        user = User.getById("make\u1000000", false);
+        assertNotNull("we do not prevent accesses to the phony name, alas", user);
+        user = User.getById("make$1000000", false);
+        assertCorrectConfig(user, "users/make$00241000000/config.xml");
+        assertFalse(new File(rootDir, "make$1000000").exists());
+        assertTrue("but asking for the real name triggers migration", user.getConfigFile().getFile().exists());
```

Comment:
![usernames](https://user-images.githubusercontent.com/154109/32622750-eb6aab44-c551-11e7-8b37-09aa854f4f40.png)

Not really fixable unless we pick a different database directory than `users`, or start storing a (usually redundant) `<id>…</id>` field, etc.

Reply:
In the first session after migration, you may see bogus usernames like this. But the config file is properly migrated as soon as you, say, log in as the real username. After the next restart, there is no trace of the bogus username `makeက000`. (Just do not get _really_ greedy and try to log in as `make$00241000000`!)